runtime/interpreter/value_range.go

```diff
@@ -0,0 +1,208 @@
+	if step.Equal(interpreter, locationRange, interpreter.GetValueForIntegerType(0, rangeType.ElementType)) {
+		panic(InclusiveRangeConstructionError{
+			LocationRange: locationRange,
+			Message:       "step value cannot be zero",
+		})
+	}
+
+	// Validate that the sequence is moving towards the end value.
+	// If start < end, step must be > 0
+	// If start > end, step must be < 0
+	// If start == end, step doesn't matter.
+	if (start.Less(interpreter, end, locationRange) && step.Less(interpreter, interpreter.GetValueForIntegerType(0, rangeType.ElementType), locationRange)) ||
+		(start.Greater(interpreter, end, locationRange) && step.Greater(interpreter, interpreter.GetValueForIntegerType(0, rangeType.ElementType), locationRange)) {
+
+		panic(InclusiveRangeConstructionError{
+			LocationRange: locationRange,
+			Message:       fmt.Sprintf("sequence is moving away from end: %s due to the value of step: %s and start: %s", end, step, start),
+		})
+	}
+
+	fields := []CompositeField{
+		{
+			Name:  sema.InclusiveRangeTypeStartFieldName,
+			Value: start,
+		},
+		{
+			Name:  sema.InclusiveRangeTypeEndFieldName,
+			Value: end,
+		},
+		{
+			Name:  sema.InclusiveRangeTypeStepFieldName,
+			Value: step,
+		},
+	}
+
+	rangeSemaType := getInclusiveRangeSemaType(interpreter, rangeType)
+
+	rangeValue := NewCompositeValueWithStaticType(
+		interpreter,
+		locationRange,
+		nil,
+		rangeSemaType.QualifiedString(),
+		common.CompositeKindStructure,
+		fields,
+		common.ZeroAddress,
+		rangeType,
+	)
+
+	rangeValue.Functions = map[string]FunctionValue{
+		sema.InclusiveRangeTypeContainsFunctionName: NewHostFunctionValue(
+			interpreter,
+			sema.InclusiveRangeContainsFunctionType(
+				rangeSemaType.MemberType,
+			),
+			func(invocation Invocation) Value {
+				needleInteger := convertAndAssertIntegerValue(invocation.Arguments[0])
+
+				return rangeContains(
+					rangeValue,
+					rangeType,
+					invocation.Interpreter,
+					invocation.LocationRange,
+					needleInteger,
+				)
+			},
+		),
+	}
+
+	return rangeValue
+}
+
+func getInclusiveRangeSemaType(interpreter *Interpreter, rangeType InclusiveRangeStaticType) *sema.InclusiveRangeType {
+	return interpreter.MustConvertStaticToSemaType(rangeType).(*sema.InclusiveRangeType)
+}
+
+func rangeContains(
+	rangeValue *CompositeValue,
+	rangeType InclusiveRangeStaticType,
+	interpreter *Interpreter,
+	locationRange LocationRange,
+	needleValue IntegerValue,
+) BoolValue {
+	start := getFieldAsIntegerValue(rangeValue, interpreter, locationRange, sema.InclusiveRangeTypeStartFieldName)
+	endInclusive := getFieldAsIntegerValue(rangeValue, interpreter, locationRange, sema.InclusiveRangeTypeEndFieldName)
+	step := getFieldAsIntegerValue(rangeValue, interpreter, locationRange, sema.InclusiveRangeTypeStepFieldName)
+
+	var result bool
+	result = start.Equal(interpreter, locationRange, needleValue) ||
+		endInclusive.Equal(interpreter, locationRange, needleValue)
+
+	if !result {
```

Comment:
maybe exit early here, by inverting the condtion
```go
if result {
    return TrueValue
}
```

Reply:
Done.